File: poker.cpp
#include <iostream>
#include <vector>
#include <utility>

class MaxHeap {
 private:
  std::vector<std::pair<std::string, double>> heap;

  int parent(int i) { return (i - 1) / 2; }

  int left(int i) { return (2 * i + 1); }

  int right(int i) { return (2 * i + 2); }

  void heapify_up(int i) {
    while (i > 0 && heap[parent(i)].second < heap[i].second) {
      std::swap(heap[parent(i)], heap[i]);
      i = parent(i);
    }
  }

  void heapify_down(int i) {
    int max_index = i;
    int l = left(i);
    int r = right(i);

    if (l < heap.size() && heap[l].second > heap[max_index].second) max_index = l;
    if (r < heap.size() && heap[r].second > heap[max_index].second) max_index = r;

    if (i != max_index) {
      std::swap(heap[i], heap[max_index]);
      heapify_down(max_index);
    }
  }

 public:
  MaxHeap() {}

  void insert(std::string key, double value) {
    heap.push_back({key, value});
    heapify_up(heap.size() - 1);
  }

  std::pair<std::string, double> get_max() {
    if (heap.size() == 0) {
      std::cout << "Heap is empty" << std::endl;
      return {"", -1};
    }
    return heap[0];
  }

  std::pair<std::string, double> extract_max() {
    if (heap.size() == 0) {
      std::cout << "Heap is empty" << std::endl;
      return {"", -1};
    }
    std::pair<std::string, double> max = heap[0];
    heap[0] = heap.back();
    heap.pop_back();
    heapify_down(0);
    return max;
  }

  void print_heap() {
    for (auto i : heap) std::cout << i.first << ":" << i.second << " ";
    std::cout << std::endl;
  }

  // print list of people who owe money
  void print_loss() {
        for (auto i : heap) {
        std::cout << i.first << " owes $" << i.second;
        std::cout << std::endl;
    }
  }

  // print list of people who earn money
  void print_gain() {
    for (auto i : heap) {
      std::cout << i.first << " earns $" << i.second;
      std::cout << std::endl;
    }
  }
};

// 
void rearrange(MaxHeap& greater, MaxHeap& less) {
    greater.insert(greater.get_max().first, greater.get_max().second - less.get_max().second);
    greater.extract_max();

    less.insert(less.get_max().first, 0);
    less.extract_max();
}


int main() {
    MaxHeap loss;
    MaxHeap gain;
    
    std::string name;
    std::string chipStr;
    std::string buyStr;
    double chips;
    double buyIn;
    std::cout << "For each player, enter name, amount of chips, and total bought in seperated by space: \n";
    for (int i = 0; i < 8; i++) {
        std::cin >> name >> chipStr >> buyStr;
        try {
          chips = std::stod(chipStr);
          buyIn = std::stod(buyStr);
        } 
        catch (const std::invalid_argument& e) {
          std::cout << "Invalid input: " << e.what() << std::endl;
        }
        double net = chips - buyIn;
        if (name == "dick") {
            net = 0;
        }
        if (net > 0) {
            gain.insert(name, net);
        }
        else if (net < 0) {
            loss.insert(name, -1 * net);
        }
    }

    std::cout << '\n';

    loss.print_loss();
    std::cout << '\n';
    gain.print_gain();

    std::cout << '\n';

    while (gain.get_max().second != 0 && loss.get_max().second != 0) {
        // if highest earner is less than biggest loser
        if (gain.get_max().second < loss.get_max().second) {
            std::cout << loss.get_max().first << " pays " << gain.get_max().first << " $" << gain.get_max().second << '\n';
            rearrange(loss, gain);
        }

        //if highest earner is greater than or equal to biggest loser
        else {
            std::cout << loss.get_max().first << " pays " << gain.get_max().first << " $" << loss.get_max().second << '\n';
            rearrange(gain, loss);
        }
        
    }

    return 0;
}

File: poker.js
class MaxHeap {
    constructor() {
      this.heap = [];
    }
  
    parent(i) {
      return Math.floor((i - 1) / 2);
    }
  
    left(i) {
      return 2 * i + 1;
    }
  
    right(i) {
      return 2 * i + 2;
    }
  
    heapify_up(i) {
      while (i > 0 && this.heap[this.parent(i)][1] < this.heap[i][1]) {
        [this.heap[this.parent(i)], this.heap[i]] = [this.heap[i], this.heap[this.parent(i)]];
        i = this.parent(i);
      }
    }
  
    heapify_down(i) {
      let max_index = i;
      let l = this.left(i);
      let r = this.right(i);
  
      if (l < this.heap.length && this.heap[l][1] > this.heap[max_index][1]) max_index = l;
      if (r < this.heap.length && this.heap[r][1] > this.heap[max_index][1]) max_index = r;
  
      if (i !== max_index) {
        [this.heap[i], this.heap[max_index]] = [this.heap[max_index], this.heap[i]];
        this.heapify_down(max_index);
      }
    }
  
    insert(key, value) {
      this.heap.push([key, value]);
      this.heapify_up(this.heap.length - 1);
    }
  
    get_max() {
      if (this.heap.length === 0) {
        console.log("Heap is empty");
        return ["", -1];
      }
      return this.heap[0];
    }
  
    extract_max() {
      if (this.heap.length === 0) {
        console.log("Heap is empty");
        return ["", -1];
      }
      let max = this.heap[0];
      this.heap[0] = this.heap.pop();
      this.heapify_down(0);
      return max;
    }
  
    print_heap() {
      console.log(this.heap.map(i => i[0] + ":" + i[1]).join(" ")); 
    }
  
    print_loss() {
      let stri = this.heap.map(i => i[0] + " owes $" + i[1]).join("\n");
      return stri;
    }
  
    print_gain() {
      let stri = this.heap.map(i => i[0] + " earns $" + i[1]).join("\n");
      return stri;
    }
  }


function rearrange(greater, less) {
    greater.insert(greater.get_max()[0], greater.get_max()[1] - less.get_max()[1]);
    greater.extract_max();
  
    less.insert(less.get_max()[0], 0);
    less.extract_max();
  }

  function main() {
    const loss = new MaxHeap();
    const gain = new MaxHeap();
  
    // // let chipStr;
    // // let buyStr;
    // console.log("For each player, enter name, amount of chips, and total bought in seperated by space: ");
    // const nameInput = document.getElementById(`nameInput${i}`);
    // const chipInput = document.getElementById(`chipInput${i}`);
    // const buyInput = document.getElementById(`buyInput${i}`);
    
    const nameArray = ["p1name", "p2name", "p3name", "p4name", "p5name", "p6name", "p7name", "p8name", ];
    const chipsArray = ["p1total", "p2total", "p3total", "p4total", "p5total", "p6total", "p7total", "p8total"];
    const buyInArray = ["p1worth", "p2worth", "p3worth", "p4worth", "p5worth", "p6worth", "p7worth", "p8worth"];
    
    

    // call to change each element
    for (let i = 0; i < 8; i++) {
        nameArray[i] = document.getElementById(nameArray[i]).value;
        chipsArray[i] = document.getElementById(chipsArray[i]).value;
        buyInArray[i] = document.getElementById(buyInArray[i]).value;
    }
    

    for (let i = 0; i < 8; i++) {
    //   try {
    //     chips = parseFloat(chipStr);
    //     buyIn = parseFloat(buyStr);
    //   } catch (e) {
    //     console.log(`Invalid input: ${e.message}`);
    //     process.exit(1);
    //   }
      let net = parseFloat(chipsArray[i]) - parseFloat(buyInArray[i]);
      if (nameArray[i] === "none") {
        net = 0;
      }
      if (net > 0) {
        gain.insert(nameArray[i], net);
      } 
      else if (net < 0) {
        loss.insert(nameArray[i], -1 * net);
      }
    }

    let summary = loss.print_loss() + "\n" + "\n" + gain.print_gain();
    console.log(summary);

    let trans = "";
    while (gain.get_max()[1] !== 0 && loss.get_max()[1] !== 0) {
      if ( gain.get_max()[1] < loss.get_max()[1]) {
        trans = trans + `${loss.get_max()[0]} pays ${gain.get_max()[0]} $${gain.get_max()[1]}` + "\n";
        rearrange(loss, gain);
      } 
      else {
        trans = trans + `${loss.get_max()[0]} pays ${ gain.get_max()[0]} $${loss.get_max()[1]}` + "\n";
        rearrange( gain, loss);
      }
    }
    console.log(trans);
  
    return 0;
  }
  
  


        